src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing/DependencyInjectionValidation.cs

```diff
@@ -0,0 +1,154 @@
+using System.Reflection;
+using Autofac;
+using Microsoft.DotNet.ServiceFabric.ServiceHost;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.Internal.DependencyInjection.Testing
+{
+    public static class DependencyInjectionValidation
+    {
+        private static readonly ImmutableList<Type> s_exemptTypes = ImmutableList.Create(
+            typeof(ServiceContext),
+            // IConfigure options is a strange type, and built into the options framework, no need to validation
+            typeof(IConfigureOptions<>),
+            // ILifetimeScope comes from Autofac, and we are only checking the ASP.NET half of this
+            typeof(ILifetimeScope),
+            typeof(MemoryCacheOptions)
+        );
+
+        private static readonly ImmutableList<string> s_exemptNamespaces = ImmutableList.Create(
+            "Microsoft.ApplicationInsights.AspNetCore"
+        );
+
+        public static bool IsDependencyResolutionCoherent(Action<ServiceCollection> register, bool includeServiceHost, out string errorMessage)
+        {
+            errorMessage = null;
+
+
+            var services = new ServiceCollection();
+            if (includeServiceHost)
+            {
+                Environment.SetEnvironmentVariable("ENVIRONMENT", "XUNIT");
+                ServiceHost.ConfigureDefaultServices(services);
+            }
+
+            register(services);
+
+            foreach (ServiceDescriptor service in services)
+            {
+                if (service.ImplementationType == null)
+                {
+                    continue;
+                }
+
+                if (IsExemptType(service.ImplementationType) || IsExemptType(service.ServiceType))
+                {
+                    continue;
+                }
+
+                bool foundConstructable = false;
+                ConstructorInfo[] constructors = service.ImplementationType
+                    .GetConstructors(BindingFlags.Public | BindingFlags.Instance)
+                    .OrderBy(c => c.GetParameters().Length)
+                    .ToArray();
+
+                if (constructors.Length == 0)
+                {
+                    // zero constructor things are implicitly constructable
+                    continue;
+                }
+
+                foreach (ConstructorInfo ctor in constructors)
+                {
+                    bool ctorWorks = true;
+
+                    foreach (ParameterInfo p in ctor.GetParameters())
+                    {
+                        ServiceDescriptor parameterService = services.FirstOrDefault(s => IsMatchingServiceRegistration(s.ServiceType, p.ParameterType));
+                        if (parameterService != null)
+                        {
+                            continue;
+                        }
+
+                        // Save the first error message, since it's likely to be the most useful
```

Comment:
Should we save all of the error messages? These errors aren't like compiled code where the first error causes more errors. I would like to see everything I haven't registered in one go, rather than have to keep re-running the test after fixing one problem.

Reply:
I guess we could build them all up for a single constructor.  I don't want to mix 2 constructors together, since that would be confusing, But I could give you all 3 parameters that are missing for one of them. :-)